src/dmd/backend/dlist.d

```diff
@@ -0,0 +1,571 @@
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(DMDSRC backend/tk/_dlist.d)
+ */
+
+module dmd.backend.dlist;
+
+import core.stdc.stdarg;
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+
+extern (C++):
+
+nothrow @nogc
+{
+
+/* **************** TYPEDEFS AND DEFINES ****************** */
+
+struct LIST
+{
+        /* Do not access items in this struct directly, use the         */
+        /* functions designed for that purpose.                         */
+        LIST* next;             /* next element in list                 */
+        int count;              /* when 0, element may be deleted       */
+        union
+        {       void* ptr;      /* data pointer                         */
+                int data;
+        }
+}
+
+alias LIST* list_t;             /* pointer to a list entry              */
+
+/* FPNULL is a null function pointer designed to be an argument to
+ * list_free().
+ */
+
+alias void function(void*) @nogc nothrow list_free_fp;
+
+enum FPNULL = cast(list_free_fp)null;
+
+/* **************** PUBLIC VARIABLES ********************* */
+
+__gshared
+{
+    int list_inited;         // != 0 if list package is initialized
+    list_t list_freelist;
+    int nlist;
+}
+
+/* **************** PUBLIC FUNCTIONS ********************* */
+
+/********************************
+ * Create link to existing list, that is, share the list with
+ * somebody else.
+ *
+ * Returns:
+ *    pointer to that list entry.
+ */
+
+list_t list_link(list_t list)
+{
+    if (list)
+        ++list.count;
+    return list;
+}
+
+/********************************
+ * Returns:
+ *    pointer to next entry in list.
+ */
+
+list_t list_next(list_t list) { return list.next; }
+
+/********************************
+ * Returns:
+ *    ptr from list entry.
+ */
+
+void* list_ptr(list_t list) { return list.ptr; }
+
+/********************************
+ * Returns:
+ *    integer item from list entry.
+ */
+
+int list_data(list_t list) { return list.data; }
+
+/********************************
+ * Append integer item to list.
+ */
+
+void list_appenddata(list_t* plist, int d)
+{
+    list_append(plist, null).data = d;
+}
+
+/********************************
+ * Prepend integer item to list.
+ */
+
+void list_prependdata(list_t *plist,int d)
+{
+    list_prepend(plist, null).data = d;
+}
+
+/**********************
+ * Initialize list package.
+ * Output:
+ *      list_inited = 1
+ */
+
+void list_init()
+{
+    if (list_inited == 0)
+    {
+        nlist = 0;
+        list_inited++;
+    }
+}
+
+/*******************
+ * Terminate list package.
+ * Output:
+ *      list_inited = 0
+ */
+
+void list_term()
+{
+    if (list_inited)
+    {
+        debug printf("Max # of lists = %d\n",nlist);
+        while (list_freelist)
+        {
+            list_t list = list_next(list_freelist);
+            list_freelist = list;
```

Comment:
The C implementation had a call to `list_delete` here.

Reply:
Yes, it does seem to be missing!